src/tests/unit/tests/DevTools/target-page-inspector.test.ts

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+import { TargetPageInspector } from 'Devtools/target-page-inspector';
+import { IMock, It, Mock, Times } from 'typemoq';
+
+describe('TargetPageInspector', () => {
+    type InspectedWindow = typeof chrome.devtools.inspectedWindow;
+
+    let inspectedWindowMock: IMock<InspectedWindow>;
+    let testSubject: TargetPageInspector;
+
+    const testFrameUrl = 'test-frame-url';
+
+    beforeEach(() => {
+        inspectedWindowMock = Mock.ofType<InspectedWindow>();
+        testSubject = new TargetPageInspector(inspectedWindowMock.object);
+    });
+
+    const safeSelectors = [
+        '#id-selector',
+        '.class-selector',
+        'article > p',
+        'article h2',
+        'article ~ h2',
+        'article + h2',
+        'li:last-child',
+        'input:checked + label',
+        'a[target]',
+        'a[href*="login"]',
+        'a[href^="https://"]',
+        'a[href$=".pdf"]',
+        'a[rel~="tag"]',
+        'a[lang|="en"]',
+        'li:nth-child(3n)',
+        'div:not(.awesome)',
+        'div::after',
+        '#result;button',
+    ];
+
+    it.each(safeSelectors)('calls eval on the target page with selector = %s', safeSelector => {
+        const expectedScript =
+            'inspect(document.querySelector(' + JSON.stringify(safeSelector) + '))';
+
+        testSubject.inspectElement(safeSelector, testFrameUrl);
+
+        inspectedWindowMock.verify(
+            inspected => inspected.eval(expectedScript, { frameURL: testFrameUrl } as any),
+            Times.once(),
+        );
+    });
+
+    it('throws for a non string selector value', () => {
+        const unsafeSelector = { description: 'this is not a string' } as any;
+
+        const act = () => testSubject.inspectElement(unsafeSelector, testFrameUrl);
+
+        inspectedWindowMock.verify(
+            inspected => inspected.eval(It.isAny(), It.isAny()),
+            Times.never(),
+        );
+
+        expect(act).toThrow('selector is not a string');
+    });
+
+    it('properly handles quotes', () => {
+        const unsafeSelector = `body'); alert('xss`; // this is an attempt to run arbitrary js code
+
+        const expectedScript =
+            'inspect(document.querySelector(' + JSON.stringify(unsafeSelector) + '))';
+
+        testSubject.inspectElement(unsafeSelector, testFrameUrl);
+
+        inspectedWindowMock.verify(
+            inspected => inspected.eval(expectedScript, { frameURL: testFrameUrl } as any),
```

Comment:
This doesn't feel like it's testing the behavior we really care about (that it doesn't escape out of the string); I wonder if better would be to do something like:
* Replace the `alert` with a throw statement in the `unsafeSelector`
* define `inspect` and `document.querySelector` in the context of the test (as no-ops)
* Have the test literally eval the string passed to `inspected.eval`, such that the test would fail (via the `throw` in `unsafeSelector`) if the quoting weren't correct

Reply:
This is a very good suggestion! The test went from "the bare minimum" to be a very strong way to probe we are approaching this ok.

I did find some difficulties defining our dependencies in the test context: I tried to use global mocks (literally `GlobalMock`) but I encountered some weird issues from typemoq (an error with: "cannot override querySelector"). The weirdest part was, the first selector (from the array, since I'm using ìt.each` showed this error but the test passed, while the second selector both, showed the error and also the test failed. It didn't matter which selector I put first and second place. 

In the end, I chose to use a "normal" mock for the `querySelector` function which I use to override `document.querySelector` (there is an `afterEach` where I make sure the original `querySelector` is assign back to the `document` object)
In the end, I